### ISSUE TITLE
GetMeetingsInfo retrieves incorrect information from BBB

### ISSUE THREAD

**Issue by flyapen** [ICTO/BBB-Load-Balancer#16]
The getmeetingsinfo that queries if a meetings is running on the loadbalancer, does not listen well and thinks a meeting is not running. The failed getMeeetingsInfo destroys the LB memory of a meeting The meeting was running and in the frontend it shows, because of this that it is not running, while in the loadbalancer on the admin page shows as running.

**Comment by brunogoossens** [comment]
Can you give me an example so I can reproduce this problem?

**Comment by flyapen** [comment]
Not directly but the same problem has been reported by 2 different load balancer users

**Comment by kentbborjke** [comment]
We have had that error when a meeting is running a long time, suddenly in the frontend (mconf-web 2.0) the meeting is shown as not running. When looking in the loadbalancer first page (servers) it is shown as running. The only way as I see it, is to let a meeting run a long time, there is no way as far as I know to know when this happens.

**Comment by kentbborjke** [comment]
I can share an account on my mconf-web and server if someone wants to test.

**Comment by kentbborjke** [comment]
I can share an account on my mconf-web and server if someone wants to test.

**Comment by kentbborjke** [comment]
I can share an account on my mconf-web and server if someone wants to test.

**Comment by kentbborjke** [comment]
I can share an account on my mconf-web and server if someone wants to test.

**Comment by brunogoossens** [comment]
The getMeetingInfoAction function is not modifying any data from the response from the BBB server. So if the BBB returns that the meeting is not running, this response will also be returned by the BBBLB.

**Comment by kentbborjke** [comment]
Kristof
Would an interrupt in the network, make LB thinks meeting is not running?

**Comment by brunogoossens** [comment]
Network failure will return: 
````
<response>
                <returncode>FAILED</returncode>
                <messageKey>connectionError</messageKey>
                <message>could not connect to the server</message>
            </response>
````

**Comment by kentbborjke** [comment]
OK so network failure is not the problem

The end API has not been sent because the meeting has not actually ended, but the LB sends to frontend meeting is not running. As you say as I understand it, this must mean that BBB sends back that meeting is not running though it is.

Can this be a BBB Problem?

Sorry for going on but I really appreciate the LB software.

Kent

**Comment by flyapen** [comment]
Hi Kent,

Could be a BBB problem, since the LB looks up the correct BBB server, does the API call to that server and returns that response from BBB. Could you verify that when the problem occurs the getmeetingsinfo API call is called on the correct BBB server ( from the nginx access logs for instance ). If that is the case then the problem is with BBB, otherwise it might be that the LB is doing the API call on the wrong server for some reason ( and returning the correct not running response ).

**Comment by kentbborjke** [comment]
Yes Kristof I will do a longer test with two servers on the LB. I will report if the error occurs and give you some logs.

Kent

**Comment by kentbborjke** [comment]
I  give you while waiting on the error this from nginx /project_error.log, should be something from former error here maybe:

2016/03/07 07:53:53 [error] 6782#0: *132406 FastCGI sent in stderr: "PHP message: PHP Warning:  SimpleXMLElement::__construct(): Entity: line 6: parser error : Opening and ending tag mismatch: hr line 5 and body in /var/www/lb/src/BBBLoadBalancer/BBBBundle/Controller/BBBAPIController.php on line 222

PHP message: PHP Warning:  SimpleXMLElement::__construct(): &lt;/body&gt; in /var/www/lb/src/BBBLoadBalancer/BBBBundle/Controller/BBBAPIController.php on line 222

PHP message: PHP Warning:  SimpleXMLElement::__construct(): Entity: line 7: parser error : Opening and ending tag mismatch: body line 3 and html in /var/www/lb/src/BBBLoadBalancer/BBBBundle/Controller/BBBAPIController.php on line 222

PHP message: PHP Warning:  SimpleXMLElement::__construct(): Entity: line 8: parser error : Premature end of data in tag html line 1 in /var/www/lb/src/BBBLoadBalancer/BBBBundle/Controller/BBBAPIController.php on line 222" while reading response header from upstream, client: 85.227.8.132, server: lb.biblestudentsonline.org, request: "GET /bigbluebutton/api/getMeetings?checksum=125f3c17627707de1993d94625edeae22f0216be HTTP/1.1", upstream: "fastcgi://unix:/var/run/php5-fpm.sock:", host: "lb.biblestudentsonline.org:1083"

2016/03/07 07:53:54 [error] 6782#0: *132408 FastCGI sent in stderr: "PHP message: PHP Warning:  SimpleXMLElement::__construct(): Entity: line 6: parser error : Opening and ending tag mismatch: hr line 5 and body in /var/www/lb/src/BBBLoadBalancer/BBBBundle/Controller/BBBAPIController.php on line 222

PHP message: PHP Warning:  SimpleXMLElement::__construct(): Entity: line 8: parser error : Premature end of data in tag html line 1 in /var/www/lb/src/BBBLoadBalancer/BBBBundle/Controller/BBBAPIController.php on line 222" while reading response header from upstream, client: 86.35.54.16, server: lb.biblestudentsonline.org, request: "GET /bigbluebutton/api/getMeetings?checksum=125f3c17627707de1993d94625edeae22f0216be HTTP/1.1", upstream: "fastcgi://unix:/var/run/php5-fpm.sock:", host: "lb.biblestudentsonline.org:1083"

Regtards
Kent

**Comment by kentbborjke** [comment]
There is some differences in Stephens suggestion for Setconfig.xml and your suggestion I present them here

Stephens Version:

else{
            $server = $meeting->getServer();

            $newconfigXML = str_replace("_LBHOST_",$server->getName(),$configXML);
            $this->get('logger')->info("DATA:", array("newconfigXML" => $newconfigXML ) );

            $chkstring ="setConfigXMLconfigXML=".urlencode($newconfigXML)."&meetingID=".$meetingID;
            $newchecksum = sha1($chkstring.$salt);
            $this->get('logger')->info("DATA:", array("Orig Checksum" => $checksum) );
            $this->get('logger')->info("DATA:", array("New Checksum" => $newchecksum) );

            **$postUrl = $server->getUrl()."/bigbluebutton/api/setConfigXML?configXML=".urlencode($newconfigXML)."&meetingID=".$meetingID."&checksum=".$newchecksum;**
            $data = urlencode($data);
            $return = $this->get('bbb')->doPostRequest($postUrl,"hi");
               # $xml=simplexml_load_string($xml);
               # $xml = new \SimpleXMLElement($return);
               # $sxe = simplexml_load_string($return);
               # $token  = (string) $sxe->configToken[0];
               # $a = array('meetingID' =>$output['meetingID'],'configXML' => $data,'token'=>$token);
               # $collection->insert($a);
                $response = new Response($return);
        }
        $response->headers->set('Content-Type', 'text/xml');
        return $response;
    }

-------------------------

Your Version here:

else {
            $server = $meeting->getServer();
            $newconfigXML = str_replace("_LBHOST_", $server->getName(), $configXML);
            $this->get('logger')->debug("DATA:", array("newconfigXML" => $newconfigXML));
          **  $data = "configXML=" . urlencode($newconfigXML) . "&meetingID=" . $meetingID;**
            $chkstring ="setConfigXML" . $data;
            $newchecksum = sha1($chkstring . $salt);
            $this->get('logger')->debug("DATA:", array("Orig Checksum" => $checksum));
            $this->get('logger')->debug("DATA:", array("New Checksum" => $newchecksum));
            $postUrl = $server->getUrl() . "/bigbluebutton/api/setConfigXML?" . $data . "&checksum=" . $newchecksum;
            $return = $this->get('bbb')->doPostRequest($postUrl, $data);
            $response = new Response($return);
        }
        $response->headers->set('Content-Type', 'text/xml');
        return $response;
    }

--------------------------

The lines that differ I have bolded

Regards
Kent

**Comment by kentbborjke** [comment]
I changed crontab -e from default which is ***** to check every 2 minutes and clean every 4:th minute. 
It seems as if this has had a good influence since I have ran tests of 2 to 4 meetings during at least all together 25 hours and no error appeared. 

Of course this does not mean the error is absolutely gone but no error so far.

Regards
Kent

**Comment by kentbborjke** [comment]
Today this morning two servers went down from the frontend at the same time as before in teh morning 4 to 5 a clock. There are meeting in the us at that time.

This afternoon it happened here in Sweden while testing with two meetings. At the same time the meetings disappeared from the frontend (mconf-web) i had a ping warning from nagios to that server.

So it seems that it is some very brief network reason to the fact that the LB misses a meeting.

Regards
Kent

**Comment by flyapen** [comment]
Hi Kent,

That would explain why the behaviours changes when changing the crontab. It might be that due to some network issue during the time when the crontab runs a server gets marked as offline. Any requests to the load balancer during that brief period would send the request to the "wrong" server resulting in the meeting not running response.

I'm not quite sure what happens when the server is marked online and a request is performed to the BBB server that is unreachable due to a network issue. I would expect that an error is returned or the next server is contacted, again resulting in that meeting not running response.

Maybe it's an idea to continue to monitor this  to see if this is the cause of the issue. And not some coincidence.

**Comment by kentbborjke** [comment]
Yes I will install smoke ping on both my servers we will see what comes out of that.

Regards
Kent

**Comment by kentbborjke** [comment]
I have measured ping with smoke-ping and at the same time running two meetings on different servers connected to the loadbalancer through the frontend (mconf-web).

When the frontend lost a meeting I saw a ping one one server that had a top value, but the thing is it was on the other server where the meeting that was lost did not run.

On the server where the meeting ran I saw not any strange values.

I give you the images from smokeping here:

![sthlmloss](https://cloud.githubusercontent.com/assets/6828340/13898367/268def4e-edcf-11e5-8915-0fe209afc5bd.png)
![smokepingsthlm](https://cloud.githubusercontent.com/assets/6828340/13898368/268dfb60-edcf-11e5-8896-8cdf204f83d1.png)
![smokepingus](https://cloud.githubusercontent.com/assets/6828340/13898380/7f556ff8-edcf-11e5-8afd-b3045045be9e.png)

My thinking is that maybe the LB did go to the wrong server to check.

Another thing I found out is that the frontend checks meetings through api calls and it never fails, I have used mconf-web for a couple of years and never dropped a meeting.

Regards
Kent

**Comment by kentbborjke** [comment]
A new meeting was lost at the same time as the LB had a little top in ping measure Here is the pic to show the graph from when the meeting was lost.

![lbloss](https://cloud.githubusercontent.com/assets/6828340/13901664/2291871e-ee2b-11e5-94bc-f56ed0604dab.png)

Kent

**Comment by kentbborjke** [comment]
Hi

After doing test 48 hour in a row without error, with this settings in crontab -e:

4 * * * * /var/www/lb/app/console bbblb:servers:check --env=prod

4 * * * * /var/www/lb/app/console bbblb:meetings:cleanup --env=prod

I have started to use LB in production environment, I will let you know if we have further problems.

Regards
Kent

**Comment by kentbborjke** [comment]
The problem was partly solved with the workaround wit the crontab settings. Tonight we had a new server is down situation when the meeting could not be opened due to the fact that the lb could not **see** the servers. here is some logs from the LB prod.log:

[2016-04-06 18:21:12] app.DEBUG: Request to BBB Server {"url":"

http://xxxx.xxxxxxxxxx.com/bigbluebutton/api/join?createTime=1459984855742&fullName=obr&meetingID=923ad5a1-2f9b-496e-a768-75bfcb6aebd0-1441666961&password=db2ea886-455b-4085-adeb-c419a99ca11c&userID=21&checksum=70de20d33c2bea949f80c46c97d1c22997916d00","output":false} []

[2016-04-06 18:21:12] app.ERROR: Server did not respond. {"Server ID":"56d610319378592d638b4567","Server URL":"http://xxxx.xxxxxxxxxx.com"} []
[2016-04-06 18:21:22] app.DEBUG: Request to BBB Server {"url":"http://xxxx.xxxxxxxxxx.com/bigbluebutton/api","output":false}

[2016-04-06 20:05:33] app.DEBUG: Request to BBB Server {"url":"

http://xxxx.xxxxxxxxxx.com/bigbluebutton/api/isMeetingRunning?meetingID=923ad5a1-2f9b-496e-a768-75bfcb6aebd0-1441666961&checksum=f3b5a1dfeb9f4c47560ac26472bfe139be5b2afd","output":false} []

[2016-04-06 20:05:33] app.ERROR: Server did not respond. {"Server ID":"56d610319378592d638b4567","Server URL":"http://xxxx.xxxxxxxxxx.com"} []
[2016-04-06 20:05:38] app.DEBUG: Request to BBB Server {"url":"http://xxxx.xxxxxxxxxx.com/bigbluebutton/api","output":false} []
[2016-04-06 20:05:38] app.INFO: Server is down. {"Server_id":"56d610319378592d638b4567","Server URL":"http://xxxx.xxxxxxxxxx.com"}

I upload a pic from the frontend mconf-web that it could not start a meeting because there where no available bbb servers.

I have checked uptime-robot where i have an account that checks the servers every minute and there where no downtime on either of our servers.

It is something wrong with the way the loadbalancer reads the servers I think that makes it sometimes out of some reasons not seeing the  servers but says "servers is down" or "server did not respond".

I do not know so much about the code so I can undretsand the reason for this but I shure hope there is some solution because we do appreciate the LB.

Regards
Kent
![image531](https://cloud.githubusercontent.com/assets/6828340/14344848/44a8e1b6-fcaa-11e5-958f-1559b1fc1434.png)

**Comment by flyapen** [comment]
Hi Kent,

Strange issue, could you check the logs on your bigbluebutton server to see if the api calls are received by bigbluebutton servers ?  The relevant logs should be 

/var/log/bigbluebutton/bbb-web.log
/var/log/nginx/bigbluebutton.access.log

**Comment by kentbborjke** [comment]
Hi Kristof

Shortly after the meeting was lost and the LB could not connect to the BBB server I see this error message could it be of interest since it seems to be related to what just had happened.

http://xxxxx-xxxxxxxx/bigbluebutton/api/getMeetings?checksum=125f3c17627707de1993d94625edeae22f0216be","output":false} []
[2016-04-06 18:20:55] app.DEBUG: Request to BBB Server {"url":"http://xxxxx-xxxxxx/bigbluebutton/api","**output":false}** []
[2016-04-06 18:20:55]** app.INFO: Server is down**. {"Server_id":"56d610319378592d638b4567","Server URL":"xxxxxxxx-xxxxxxxx"} []¨

[2016-04-06 18:20:55] **php.EMERGENCY: Fatal Error: Call to a member function getUrl() on a non-object {"type":1,"file":"/var/www/lb/src/BBBLoadBalancer/BBBBundle/Controller/BBBAPIController.php","line":62,"level":6143,"stack**":[]} []

**[2016-04-06 18:20:55] request.CRITICAL: Uncaught PHP Exception Symfony\Component\Debug\Exception\FatalErrorException: "Error: Call to a member function getUrl() on a non-object" at /var/www/lb/src/BBBLoadBalancer/BBBBundle/Controller/BBBAPIController.php line 62 {"exception":"[object] (Symfony\\Component\\Debug\\Exception\\FatalErrorException(code: 0): Error: Call to a member function getUrl() on a non-object at /var/www/lb/src/BBBLoadBalancer/BBBBundle/Controller/BBBAPIController.php:62)"}** []

[2016-04-06 18:20:55] security.DEBUG: Write SecurityContext in the session [] []
[2016-04-06 18:20:57] request.INFO: Matched route "bbbloadbalancer_bbb_bbbapi_join" (parameters: "_format": "xml", "_controller": "BBBLoadBalancer\BBBBundle\Controller\BBBAPIController::joinAction", "_route": "bbbloadbalancer_bbb_bbbapi_join") [] []

Regards
Kent

**Comment by kentbborjke** [comment]
Kristof

I tested two servers without https:// and the same error occurred, of course I ran the meetings a long time in this scenario I think 8 or 9 hours.

Kent.

**Comment by kentbborjke** [comment]
Hi

After the two latest modifications, I have ran tests now, for three days with two meetings on different servers connected to the LB, no error yet. I will wait and see if and when next error comes.

Regards
Kent
